src/OpenRiaServices.DomainServices.Hosting/Framework/Services/MessageEncoders/PoxBinaryMessageEncodingBindingElement.cs

```diff
@@ -255,15 +256,224 @@ public override ArraySegment<byte> WriteMessage(Message message, int maxMessageS
+                            _position += count;
+                            _bufferWritten += count;
+                            break;
+                        }
+                        else
+                        {
+                            // Fill _buffer
+                            int toCopy = _buffer.Length - _bufferWritten;
+                            FastCopy(buffer, offset, _buffer, _bufferWritten, toCopy);
+                            _position += toCopy;
+                            offset += toCopy;
+                            count -= toCopy;
+                        }
+                    } while (count > 0);
+                }
+
+                /// <summary>
+                /// Allocate more space if buffer is full.
+                /// Ensures _buffer is non null and has space to write more bytes
+                /// </summary>
+                private void EnsureBufferCapacity()
+                {
+                    // There is space left
+                    if (_bufferWritten < _buffer.Length)
+                        return;
+
+                    // Save current buffer in list before allocating a new buffer
+                    if (_bufferList == null)
+                        _bufferList = new System.Collections.Generic.List<byte[]>(capacity: 16);
+                    _bufferList.Add(_buffer);
+                    // Ensure we never return buffer twice in case TakeBuffer below throws
+                    _buffer = null;
+
+                    _buffer = _bufferManager.TakeBuffer(Math.Min(_position, _maxSize));
+                    _bufferWritten = 0;
+                }
+
+                protected override void Dispose(bool disposing)
+                {
+                    try
+                    {
+                        base.Dispose(disposing);
+                    }
+                    finally
+                    {
+                        if (_buffer != null)
+                        {
+                            _bufferManager.ReturnBuffer(_buffer);
+                            _buffer = null;
+                        }
+
+                        if (_bufferList != null)
+                        {
+                            foreach (var buffer in _bufferList)
+                                _bufferManager.ReturnBuffer(buffer);
+                            _bufferList = null;
+                        }
+                    }
+                }
+
+                /// <summary>
+                /// Copies bytes from <paramref name="src"/> to <paramref name="dest"/> using fastes 
+                /// copy based on process bitness (x86 / x64)
+                /// </summary>
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                private static unsafe void FastCopy(byte[] src, int srcOffset, byte[] dest, int destOffset, int count)
+                {
+                    if (count == 0)
+                        return;
+
+                    if (Is64BitProcess)
```

Comment:
Blockcopy is faster for sizes above 2048 on net framework for x64, but it may not be significant enough to change the code